inception-ui-kb/src/main/java/de/tudarmstadt/ukp/inception/ui/kb/feature/FactLinkingServiceImpl.java

```diff
@@ -0,0 +1,184 @@
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package de.tudarmstadt.ukp.inception.ui.kb.feature;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.LinkedList;
+import java.util.List;
+import javax.annotation.Resource;
+
+import org.apache.uima.cas.CASException;
+import org.apache.uima.cas.text.AnnotationFS;
+import org.apache.uima.jcas.JCas;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Component;
+
+import de.tudarmstadt.ukp.clarin.webanno.api.AnnotationSchemaService;
+import de.tudarmstadt.ukp.clarin.webanno.api.annotation.action.AnnotationActionHandler;
+import de.tudarmstadt.ukp.clarin.webanno.api.annotation.model.AnnotatorState;
+import de.tudarmstadt.ukp.clarin.webanno.api.annotation.model.LinkWithRoleModel;
+import de.tudarmstadt.ukp.clarin.webanno.api.annotation.util.WebAnnoCasUtil;
+import de.tudarmstadt.ukp.clarin.webanno.model.AnnotationFeature;
+import de.tudarmstadt.ukp.clarin.webanno.model.AnnotationLayer;
+import de.tudarmstadt.ukp.clarin.webanno.model.Project;
+import de.tudarmstadt.ukp.inception.kb.KnowledgeBaseService;
+import de.tudarmstadt.ukp.inception.kb.graph.KBHandle;
+import de.tudarmstadt.ukp.inception.kb.graph.KBStatement;
+import de.tudarmstadt.ukp.inception.kb.model.KnowledgeBase;
+
+@Component(FactLinkingService.SERVICE_NAME)
+public class FactLinkingServiceImpl implements FactLinkingService
+{
+    @Resource private KnowledgeBaseService kbService;
+    @Resource private AnnotationSchemaService annotationService;
+
+    private static final String FACT_LAYER = "webanno.custom.Fact";
+    private final Logger log = LoggerFactory.getLogger(getClass());
+
+    @Override
+    public KnowledgeBase getKBByKBHandle(KBHandle kbHandle, Project aProject)
+    {
+        for (KnowledgeBase kb : kbService.getKnowledgeBases(aProject)) {
+            if (kbService.listProperties(kb, false).contains(kbHandle)) {
+                return kb;
+            }
+            if (kbService.listConcepts(kb, false).contains(kbHandle)) {
+                return kb;
+            }
+            for (KBHandle concept : kbService.listConcepts(kb, false)) {
+                if (kbService.listInstances(kb, concept.getIdentifier(), false)
+                    .contains(kbHandle)) {
+                    return kb;
+                }
+            }
+
+        }
+        return null;
+    }
+
+    @Override
+    public List<KBHandle> getKBConceptsAndInstances(Project aProject)
+    {
+        List<KBHandle> handles = new LinkedList<>();
```

Comment:
I would write Sparql for that. Also, it will break for larger, e.g. remote knowledge bases.

Reply:
Also use ArrayList, not LinkedList